### ISSUE TITLE
Requested task count does not match running allocations

### ISSUE THREAD

**Issue by rickardrosen** [hashicorp/nomad#2710]
### Nomad version
0.5.6

### Operating system and Environment details
Ubuntu 16.10

### Issue
We are seeing an issue with jobs that the count requested in the job spec does not match what is running.
Here is the output from a job with a single taskgroup with count = 2, containing a single task:
```
$ docker ps | grep customer-news
c79d35b7d12a        299477808f9c        "/opt/containerpil..."   38 minutes ago      Up 38 minutes                           customer-newsletter-web-dec83193-cd0f-b836-84a0-fb6024f0f2af
c8c9b3d3600e        299477808f9c        "/opt/containerpil..."   38 minutes ago      Up 38 minutes                           customer-newsletter-web-bcfeaea2-18b8-ec82-f1a8-6508dfd3b05c
855d30e7896b        299477808f9c        "/opt/containerpil..."   39 minutes ago      Up 39 minutes                           customer-newsletter-web-e6c2de43-bf14-bd25-97a0-c3854ac56094
3f8e6627db40        299477808f9c        "/opt/containerpil..."   39 minutes ago      Up 39 minutes                           customer-newsletter-web-e18019fa-76dc-6569-eb80-13b1df59d5db

root@xxx:~# nomad status customer-newsletter-web:uat
ID            = customer-newsletter-web:uat
Name          = customer-newsletter-web
Type          = service
Priority      = 50
Datacenters   = dc1
Status        = running
Periodic      = false
Parameterized = false

Summary
Task Group               Queued  Starting  Running  Failed  Complete  Lost
customer-newsletter-web  0       0         4        6       24        2

Allocations
ID        Eval ID   Node ID   Task Group               Desired  Status    Created At
dec83193  6cb6a002  4219f4a5  customer-newsletter-web  run      running   06/16/17 10:20:58 CEST
bcfeaea2  c56715ab  4219f4a5  customer-newsletter-web  run      running   06/16/17 10:20:28 CEST
e6c2de43  443b2f80  4219f4a5  customer-newsletter-web  run      running   06/16/17 10:19:58 CEST
e18019fa  c991f6ad  4219f4a5  customer-newsletter-web  run      running   06/16/17 10:19:28 CEST
```

Job plan for this job (single group, count 2, single task)

```
{
  "Annotations": {
    "DesiredTGUpdates": {
      "customer-newsletter-web": {
        "Ignore": 0,
        "Place": 0,
        "Migrate": 0,
        "Stop": 0,
        "InPlaceUpdate": 4,
        "DestructiveUpdate": 0
      }
    }
  },
  "FailedTGAllocs": null,
  "JobModifyIndex": 1451144,
  "CreatedEvals": null,
  "Diff": null,
  "NextPeriodicLaunch": "0001-01-01T00:00:00Z",
  "Index": 1451144
}

```

### Reproduction steps
n/a

### Job file (if appropriate)
```
{
    "Job": {
  "Region": "nordic",
  "ID": "customer-newsletter-web:uat",
  "ParentID": "",
  "Name": "customer-newsletter-web",
  "Type": "service",
  "Priority": 50,
  "AllAtOnce": false,
  "Datacenters": [
    "dc1"
  ],
  "Constraints": [
    {
      "LTarget": "${node.class}",
      "RTarget": "uat",
      "Operand": "="
    }
  ],
  "TaskGroups": [
    {
      "Name": "customer-newsletter-web",
      "Count": 2,
      "Constraints": null,
      "RestartPolicy": {
        "Attempts": 10,
        "Interval": 300000000000,
        "Delay": 25000000000,
        "Mode": "delay"
      },
      "Tasks": [
        {
          "Name": "customer-newsletter-web",
          "Driver": "docker",
          "User": "",
          "Config": {
            "image": "web-docker.domain.com/customer-newsletter-web:ddf9108",
            "command": "/opt/containerpilot/tools/containerpilot",
            "args": [
              "node",
              "webserver.js"
            ],
            "volumes": [
              "/opt/containerpilot:/opt/containerpilot"
            ],
            "labels": [
              {
                "service": "customer-newsletter-web",
                "scope": "uat",
                "owner": "customer",
                "image": "customer-newsletter-web",
                "image_tag": "ddf9108"
              }
            ],
            "logging": [
              {
                "config": [
                  {
                    "gelf-address": "udp://logging.domain.com:12203",
                    "labels": "owner,service"
                  }
                ],
                "type": "gelf"
              }
            ],
            "network_mode": "containernet",
            "dns_search_domains": [
              "domain.com"
            ],
            "port_map": []
          },
          "Env": {
            "SCOPE": "uat",
            "NOMAD_NODE_HOSTNAME": "${attr.unique.hostname}.domain.com",
            "HEALTH_CHECK_URI": "/health",
            "CONTAINERPILOT": "file:///local/containerpilot.json",
            "SERVICE_PORT": "80"
          },
          "Services": [],
          "Vault": null,
          "Templates": [],
          "Constraints": null,
          "Resources": {
            "CPU": 250,
            "MemoryMB": 128,
            "DiskMB": 0,
            "IOPS": 0,
            "Networks": [
              {
                "Device": "",
                "CIDR": "",
                "IP": "",
                "MBits": 10,
                "ReservedPorts": null,
                "DynamicPorts": null
              }
            ]
          },
          "DispatchPayload": null,
          "Meta": {
            "uuid": "f262affd-fbd9-4875-b79f-788d39979511"
          },
          "KillTimeout": 15000000000,
          "LogConfig": {
            "MaxFiles": 5,
            "MaxFileSizeMB": 10
          },
          "Artifacts": [],
          "Leader": false
        }
      ],
      "EphemeralDisk": {
        "Sticky": false,
        "SizeMB": 300,
        "Migrate": false
      },
      "Meta": null
    }
  ],
  "Update": {
    "Stagger": 30000000000,
    "MaxParallel": 1
  }
}
}
```

**Comment by schmichael** [comment]
Two things stand out to me in the output you posted:

1. Two lost allocs (was your cluster unhealthy for a time? We should recover properly, but it could be related.)
2. The second copy of each alloc created exactly 1 minute later. Not sure the significance (if any) of that yet.

Can you post any relevant logs for those running allocations? Github allows attaching files if that helps, or feel free to email me directly if you don't want to redact sensitive information.

Can you post `nomad eval-status` for the 4 evaluations that produced those 4 allocations?

Thanks!

**Comment by rickardrosen** [comment]
You are correct regarding the health and I'm pretty sure this is somehow related.
We suffered several network outages a few weeks ago.

I'll drop you some logs tonight.

Thank you!

**Comment by tantra35** [comment]
In reference issue we place our nomad `eval-status`

**Comment by rickardrosen** [comment]
So, the team owning the app did a stop and a re-deploy of the job and now all looks good with the allocations.
Unfortunately this means I've not been able to pick up the eval-status output.

It is obvious that strange things started to happen with running jobs after our cluster split/outage.

We also saw at least one job missing containers; count two, but only one allocation running.
Increase count to 3, plan/schedule, to end up with two running instances.
These jobs were stopped and restarted immediately however, and during the chaos that took place during the outage, no proper focus was given this Nomad issue.

Perhaps @tantra35 is better positioned to gather relevant logs in #2712?

**Comment by dadgar** [comment]
https://github.com/hashicorp/nomad/issues/2511

**Comment by dadgar** [comment]
Hey sorry you ran into this. We haven't quite been able to reproduce this so any more hints would be appreciated but obviously it is hard to tell what happened.

As for fixing this, Nomad 0.6.0 is largely redoing the way reconciliation happens in the scheduler. So I believe this will be fixed in 0.6.0! I'd like to keep https://github.com/hashicorp/nomad/issues/2511 open till 0.6.0 and then see if there are any more reports of this with the new version!

Thanks,
Alex
